File: packages/node/runtime/src/binary.js
import { statSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { createHash } from "node:crypto";
import { WreError, ErrorKind } from "./errors.js";

const TRIPLES = {
  "darwin:arm64": "aarch64-apple-darwin",
  "darwin:x64": "x86_64-apple-darwin",
  "linux:x64": "x86_64-unknown-linux-gnu",
  "linux:arm64": "aarch64-unknown-linux-gnu",
  "win32:x64": "x86_64-pc-windows-msvc",
};

export function currentTriple() {
  const key = `${process.platform}:${process.arch}`;
  const triple = TRIPLES[key];
  if (!triple) {
    throw new WreError(
      ErrorKind.Unsupported,
      `no known binary triple for platform "${process.platform}" arch "${process.arch}"`,
    );
  }
  return triple;
}

export function verifySha256(path, expected) {
  const data = readFileSync(path);
  const actual = createHash("sha256").update(data).digest("hex");
  if (actual !== expected) {
    throw new WreError(
      ErrorKind.Resource,
      `sha256 mismatch for "${path}": expected ${expected}, got ${actual}`,
    );
  }
}

function isFile(path) {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

export function resolveBinary(options = {}) {
  const tried = [];

  const envBinary = process.env.WRE_BINARY;
  if (envBinary) {
    tried.push(`WRE_BINARY=${envBinary}`);
    if (isFile(envBinary)) {
      return envBinary;
    }
  }

  if (options.embedded) {
    const exeName = process.platform === "win32" ? "wred.exe" : "wred";
    const candidate = join(options.embedded, exeName);
    tried.push(candidate);
    if (isFile(candidate)) {
      if (options.sha256) {
        verifySha256(candidate, options.sha256);
      }
      return candidate;
    }
  }

  const attempted = tried.length > 0 ? tried.join(", ") : "nothing (no WRE_BINARY and no embedded path given)";
  throw new WreError(ErrorKind.Resource, `could not resolve a wred binary, tried: ${attempted}`);
}

File: packages/python/wre_runtime/binary.py
from __future__ import annotations

import hashlib
import os
import platform
import sys
from typing import Dict, List, Optional, Tuple

from .errors import ResourceError, Unsupported

_TRIPLES: Dict[Tuple[str, str], str] = {
    ("darwin", "arm64"): "aarch64-apple-darwin",
    ("darwin", "x86_64"): "x86_64-apple-darwin",
    ("linux", "x86_64"): "x86_64-unknown-linux-gnu",
    ("linux", "arm64"): "aarch64-unknown-linux-gnu",
    ("win32", "x86_64"): "x86_64-pc-windows-msvc",
}


def current_triple() -> str:
    plat = sys.platform
    machine = platform.machine()
    normalized = machine
    if machine in ("arm64", "aarch64"):
        normalized = "arm64"
    elif machine in ("x86_64", "AMD64"):
        normalized = "x86_64"
    triple = _TRIPLES.get((plat, normalized))
    if triple is None:
        raise Unsupported(f"no wred binary for platform {plat!r} machine {machine!r}")
    return triple


def verify_sha256(path: str, expected: str) -> None:
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    actual = digest.hexdigest()
    if actual.lower() != expected.lower():
        raise ResourceError(
            f"sha256 mismatch for {path}: expected {expected}, got {actual}",
            retryable=True,
        )


def resolve_binary(package_dir: Optional[str] = None, sha256: Optional[str] = None) -> str:
    tried: List[str] = []
    env_path = os.environ.get("WRE_BINARY")
    if env_path:
        tried.append(env_path)
        if os.path.isabs(env_path) and os.path.isfile(env_path):
            return env_path
    if package_dir is not None:
        triple = current_triple()
        binary_name = "wred.exe" if sys.platform == "win32" else "wred"
        candidate = os.path.join(package_dir, "bin", triple, binary_name)
        tried.append(candidate)
        if os.path.isfile(candidate):
            if sha256 is not None:
                verify_sha256(candidate, sha256)
            return candidate
    listed = ", ".join(tried) if tried else "(none)"
    raise ResourceError(f"could not resolve a wred binary, tried: {listed}", retryable=True)


def cache_root() -> str:
    override = os.environ.get("WRE_CACHE_DIR")
    if override:
        return override
    xdg = os.environ.get("XDG_CACHE_HOME")
    if xdg:
        return os.path.join(xdg, "wre")
    if sys.platform == "win32":
        local_app_data = os.environ.get("LOCALAPPDATA")
        if local_app_data:
            return os.path.join(local_app_data, "wre")
        return os.path.join(os.path.expanduser("~"), "AppData", "Local", "wre")
    return os.path.join(os.path.expanduser("~"), ".cache", "wre")
